prober/http.go

```diff
@@ -65,9 +67,76 @@ func matchRegularExpressions(reader io.Reader, httpConfig config.HTTPProbe, logg
 	return true
 }
 
+// trace holds timings for a single HTTP roundtrip.
+type roundTripTrace struct {
+	tls           bool
+	start         time.Time
+	dnsDone       time.Time
+	connectDone   time.Time
+	gotConn       time.Time
+	responseStart time.Time
+	end           time.Time
+}
+
+// transport is a custom transport keeping traces for each HTTP roundtrip.
+type transport struct {
+	Transport http.RoundTripper
+	logger    log.Logger
+	traces    []*roundTripTrace
+	current   *roundTripTrace
+}
+
+func newTransport(rt http.RoundTripper, logger log.Logger) *transport {
+	return &transport{
+		Transport: rt,
+		logger:    logger,
+		traces:    []*roundTripTrace{},
+	}
+}
+
+// RoundTrip switches to a new trace, then runs embedded RoundTripper.
+func (t *transport) RoundTrip(req *http.Request) (*http.Response, error) {
+	trace := &roundTripTrace{}
+	if req.URL.Scheme == "https" {
+		trace.tls = true
+	}
+	t.current = trace
+	t.traces = append(t.traces, trace)
+	defer func() { trace.end = time.Now() }()
+	return t.Transport.RoundTrip(req)
+}
+
+func (t *transport) DNSStart(_ httptrace.DNSStartInfo) {
+	t.current.start = time.Now()
+}
+func (t *transport) DNSDone(_ httptrace.DNSDoneInfo) {
+	t.current.dnsDone = time.Now()
+}
+func (ts *transport) ConnectStart(_, _ string) {
+	t := ts.current
+	// No DNS resolution, e.g. connecting to IP directly.
+	if t.dnsDone.IsZero() {
+		t.start = time.Now()
+		t.dnsDone = t.start
+	}
+}
+func (t *transport) ConnectDone(net, addr string, err error) {
+	t.current.connectDone = time.Now()
+}
+func (t *transport) GotConn(_ httptrace.GotConnInfo) {
+	t.current.gotConn = time.Now()
+}
+func (t *transport) GotFirstResponseByte() {
+	t.current.responseStart = time.Now()
+}
+
 func ProbeHTTP(ctx context.Context, target string, module config.Module, registry *prometheus.Registry, logger log.Logger) (success bool) {
 	var redirects int
 	var (
+		durationGaugeVec = prometheus.NewGaugeVec(prometheus.GaugeOpts{
```

Comment:
We should initialise this with all the potential phases.

Reply:
How?

Reply:
Call WithLabelValues with all the potential values.

Reply:
Like this?